File: Baekjoon/7562_나이트의 이동/B_7562.py
from collections import deque


def bfs(deq):
    y_axis = [2, 1, -1, -2, -2, -1, 1, 2]
    x_axis = [1, 2, 2, 1, -1, -2, -2, -1]
    while deq:
        y, x = deq.popleft()
        for j in range(8):
            y_location = y + y_axis[j]
            x_location = x + x_axis[j]
            if 0 <= y_location < length and 0 <= x_location < length:
                if y_location == to_move_y and x_location == to_move_x:
                    return chessboard[y][x]
                if not chessboard[y_location][x_location]:
                    chessboard[y_location][x_location] = chessboard[y][x] + 1
                    deq.append([y_location, x_location])


T = int(input())
for _ in range(T):
    length = int(input())
    chessboard = [[0] * length for _ in range(length)]
    cur_y, cur_x = map(int, input().split())
    to_move_y, to_move_x = map(int, input().split())
    chessboard[cur_y][cur_x] = 1
    d = deque()
    d.append([cur_y, cur_x])
    if cur_y == to_move_y and cur_x == to_move_x:
        print(0)
    else:
        print(bfs(d))

File: Baekjoon/7562_나이트의 이동/B_7562.js
const fs = require('fs')
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n')

const T = parseInt(input[0])
let idx = 1
for (let t = 0; t < T; t++) {

    const Node = function(value, next = null) {
        this.value = value
        this.next = next
    }
    
    const Queue = function() {
        this.head = new Node(null)
        this.tail = this.head
    }

    Queue.prototype.isEmpty = function() {
        return this.head.value === null && this.tail.value === null
    }

    Queue.prototype.enQueue = function(value) {
        if (this.isEmpty()) {
            this.head = new Node(value)
            this.tail = this.head
        } else {
            const tail = this.tail
            this.tail = new Node(value)
            tail.next = this.tail
        }
    }
    
    Queue.prototype.deQueue = function() {
        if (this.isEmpty() === false) {
            const popNumber = this.head.value
            
            if (this.head === this.tail) {
                this.head = new Node(null)
                this.tail = this.head
            } else {
                this.head = this.head.next
            }

            return popNumber
        }
    }

    const bfs = function() {
        const q = new Queue()
        q.enQueue([startY, startX])
        chessBoard[startY][startX] = 1

        const yAxis = [-2, -1, 1, 2, 2, 1, -1, -2]
        const xAxis = [1, 2, 2, 1, -1, -2, -2, -1]

        while (q.isEmpty() === false) {
            const [y, x] = q.deQueue()
            
            if (y === endY && x === endX) {
                return chessBoard[endY][endX] - 1
            }
            
            for (let i = 0; i < 8; i++) {
                const yWay = y + yAxis[i]
                const xWay = x + xAxis[i]

                if (0 <= yWay && yWay < N && 0 <= xWay && xWay < N) {
                    if (chessBoard[yWay][xWay] === 0) {
                        chessBoard[yWay][xWay] = chessBoard[y][x] + 1
                        q.enQueue([yWay, xWay])
                    }
                }
            }
        }
    }

    const N = parseInt(input[idx])
    const [startY, startX] = input[idx + 1].split(' ').map(v => parseInt(v))
    const [endY, endX] = input[idx + 2].split(' ').map(v => parseInt(v))
    
    let chessBoard = new Array(N).fill(0).map(v => new Array(N).fill(0))
    const answer = bfs()
    idx += 3
    console.log(answer)
}